File: Spectrum and simulation/src/matlab/BPSK and WSS.m
% BPSK Spectrum Analysis and WSS Verification
clear all; close all; clc;

% Parameters
Fs = 1000;          % Sampling frequency
T = 0.01;           % Symbol period (100 Hz symbol rate)
A = 1;              % Amplitude
num_symbols = 10000; % Increase for better PSD estimation
samples_per_symbol = T * Fs; % Samples per symbol

% Generate random binary sequence
bits = randi([0 1], 1, num_symbols);

% BPSK modulation (0 -> +A, 1 -> -A)
symbols = A * (1 - 2 * bits);

% Create rectangular pulse shaped signal 
tx_signal = repelem(symbols, samples_per_symbol);

% Calculate theoretical PSD
f = linspace(-Fs/2, Fs/2, 1024);
theoretical_psd = (A^2 * T) * (sinc(f * T)).^2;

% Estimate practical PSD using Welch method
window_length = min(1024, length(tx_signal)/4);
[psd_est, f_est] = pwelch(tx_signal, window_length, [], 1024, Fs, 'centered');

% Convert to dB for plotting 
psd_est_db = 10*log10(psd_est + eps);
theoretical_psd_db = 10*log10(theoretical_psd + eps);

% Normalize for better comparison (optional)
psd_est_db = psd_est_db - max(psd_est_db);
theoretical_psd_db = theoretical_psd_db - max(theoretical_psd_db);

% Plotting
figure;

subplot(3,1,1);
stem(bits(1:20), 'filled');
title('First 20 Binary Bits');
xlabel('Bit Index'); ylabel('Bit Value'); grid on;

subplot(3,1,2);
plot((0:199)/Fs, tx_signal(1:200));
title('BPSK Modulated Signal (First 200 ms)');
xlabel('Time (s)'); ylabel('Amplitude'); grid on;

subplot(3,1,3);
plot(f, theoretical_psd_db, 'r--', 'LineWidth', 2); hold on;
plot(f_est, psd_est_db, 'b', 'LineWidth', 1);
title('Power Spectral Density Comparison');
xlabel('Frequency (Hz)'); ylabel('Normalized PSD (dB)');
legend('Theoretical', 'Estimated'); grid on;
xlim([-200, 200]); % Zoom around main lobe

% WSS Verification - Improved
segment_length = 1000;
num_segments = floor(length(tx_signal) / segment_length);
signal_means = zeros(1, num_segments);
signal_vars = zeros(1, num_segments);

for i = 1:num_segments
    start_idx = (i-1)*segment_length + 1;
    end_idx = i*segment_length;
    segment = tx_signal(start_idx:end_idx);
    signal_means(i) = mean(segment);
    signal_vars(i) = var(segment);
end

figure;
subplot(2,1,1);
plot((1:num_segments)*segment_length/Fs, signal_means);
title('Mean Value over Time (WSS Check)');
xlabel('Time (s)'); ylabel('Mean'); grid on;

subplot(2,1,2);
plot((1:num_segments)*segment_length/Fs, signal_vars);
title('Variance over Time (WSS Check)');
xlabel('Time (s)'); ylabel('Variance'); grid on;

% Calculate correlation to show it depends only on time difference
max_lag = 100;
[corr_vals, lags] = xcorr(tx_signal(1:1000), max_lag, 'coeff');
figure;
stem(lags/Fs, corr_vals);
title('Autocorrelation Function (Should depend only on lag)');
xlabel('Lag (s)'); ylabel('Correlation'); grid on;

File: Spectrum and simulation/src/python/BPSK and WSS.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

# Parameters
Fs = 1000          # Sampling frequency
T = 0.01           # Symbol period
A = 1              # Amplitude
num_symbols = 1000  # Number of symbols to generate
samples_per_symbol = int(T*Fs)  # Samples per symbol

# Generate random binary sequence
bits = np.random.randint(0, 2, num_symbols)

# BPSK modulation (0 -> +A, 1 -> -A)
symbols = A*(1 - 2*bits)

# Upsample and create rectangular pulse shaping
upsampled = np.repeat(symbols, samples_per_symbol)
pulse_shape = np.ones(samples_per_symbol)
tx_signal = np.convolve(upsampled, pulse_shape, mode='same')

# Calculate theoretical PSD
f = np.linspace(-Fs/2, Fs/2, 1024)
theoretical_psd = (A**2 * T) * (np.sinc(f*T))**2

# Estimate practical PSD
f_est, psd_est = signal.welch(tx_signal, Fs, nperseg=1024, scaling='density', center=True)

# Plotting
plt.figure(figsize=(12, 10))

plt.subplot(3,1,1)
plt.stem(bits[:20])
plt.title('First 20 Binary Bits')
plt.xlabel('Bit Index')
plt.ylabel('Bit Value')
plt.grid(True)

plt.subplot(3,1,2)
plt.plot(tx_signal[:200])
plt.title('BPSK Modulated Signal (First 200 Samples)')
plt.xlabel('Sample Index')
plt.ylabel('Amplitude')
plt.grid(True)

plt.subplot(3,1,3)
plt.plot(f, 10*np.log10(theoretical_psd/np.max(theoretical_psd)), 'r--', label='Theoretical')
plt.plot(f_est, 10*np.log10(psd_est/np.max(psd_est)), 'b', label='Estimated')
plt.title('Power Spectral Density')
plt.xlabel('Frequency (Hz)')
plt.ylabel('Power/Frequency (dB/Hz)')
plt.legend()
plt.grid(True)

# WSS Verification
time_instances = np.arange(1000)
signal_means = np.array([np.mean(tx_signal[i:i+1000]) for i in time_instances])

plt.figure()
plt.plot(time_instances, signal_means)
plt.title('Mean Value over Time (WSS Check)')
plt.xlabel('Time Instance')
plt.ylabel('Mean Value')
plt.grid(True)

plt.tight_layout()
plt.show()